packages/cid-verifier/src/verification.js

```diff
@@ -0,0 +1,130 @@
+ */
+async function getGoogleMalwareResult (cid, env) {
+  const datastore = env.GOOGLE_MALWARE_RESULTS
+  if (!datastore) {
+    return undefined
+  }
+
+  return await pRetry(
+    // TODO: in theory we should check each subcomponent of the pathname also.
+    // https://github.com/nftstorage/nft.storage/issues/1737
+    () => datastore.get(cid),
+    { retries: 5 }
+  )
+}
+
+/**
+ * @param {Array<string>} params
+ * @param {(params: Array<string>, request: Request, env: import('./env').Env) => Promise<Response>} fn
+ * @returns {import('itty-router').RouteHandler<Request>}
+ */
+function withRequiredQueryParams (params, fn) {
+  /**
+   * @param {Request} request
+   * @param {import('./env').Env} env
+   */
+  return async function (request, env) {
+    const searchParams = (new URL(request.url)).searchParams
+
+    for (const param of params) {
+      if (!searchParams.get(param)) {
+        return new Response(`${param} is a required query param`, { status: 400 })
+      }
+    }
+
+    return await fn(params.map(param => String(searchParams.get(param))), request, env)
+  }
+}
+
+export const verificationGet = withRequiredQueryParams(['cid'],
+  /**
+   * Returns google malware result.
+   */
+  async function (params, request, env) {
+    const [cid] = params
+
+    const denyListResource = await getFromDenyList(cid, env)
+    if (denyListResource) {
+      const { status } = JSON.parse(denyListResource)
+      if (status === 451) {
+        return new Response('BLOCKED FOR LEGAL REASONS', { status: 451 })
+      } else {
+        return new Response('MALWARE DETECTED', { status: 403 })
+      }
+    }
+
+    const googleMalwareResults = await getGoogleMalwareResult(cid, env)
+    if (!googleMalwareResults) {
+      return new Response('Verification Not Known', { status: 404 })
+    }
+
+    if (googleMalwareResults === '"PENDING"') {
+      return new Response('Verification Pending', { status: 202 })
+    }
+
+    const parsedGoogleMalwareResults = JSON.parse(googleMalwareResults)
+    if (!Object.keys(parsedGoogleMalwareResults).length) {
+      return new Response('Malware Not Detected', { status: 200 })
+    } else {
+      return new Response('MALWARE DETECTED', { status: 403 })
+    }
+  }
+)
+
+/**
+ * Process CID with malware verification parties.
+ */
+export const verificationPost = withRequiredQueryParams(['cid', 'url'],
+  async function verificationPost (params, request, env) {
+    const [cid, url] = params
+    const googleMalwareResults = await getGoogleMalwareResult(cid, env)
+
+    if (googleMalwareResults === '"PENDING"') {
+      return new Response('Verification Pending', { status: 202 })
+    }
+
+    if (googleMalwareResults) {
+      return new Response('cid already processed', { status: 409 })
+    }
+
+    // cid not processed so we process it
+
+    // we set the status to pending to limit the number of times we call google
+    env.GOOGLE_MALWARE_RESULTS.put(cid, '"PENDING"')
```

Comment:
```suggestion
    env.GOOGLE_MALWARE_RESULTS.put(cid, '"PENDING"')
```

I am concerned about KV nature being eventually consistent here. For example, if two concurrent requests happen (one in Europe and one in the USA) there is a possibility of:
1. req1.get()
2. req2.get()
3. req1.put('PENDING')
4. req1.put(result)
5. req2.put('PENDING')
6. req2.put(result)

Eventually it will be fine in the end. But building on my other comment https://github.com/web3-storage/reads/pull/36#discussion_r960765098 `packages/cid-verifier/src/bindings.d.ts`. Probably, we can go even a step further and write a key lock: `cid/google-evaluate-api.lock` with PENDING value

Once we have final value, we write to key `cid/google-evaluate-api` and delete lock. This way, we never have issues with concurrent writes in same key.

Of course this means above we should use KV.list instead of KV.get.

Let me know your thoughts here, perhaps you have other ideas in mind

Reply:
i kinda feel eventual consistency is safer and potentially serves up more concurrent requests compared to locking.

Reply:
I don't agree with safer. What if response from Google fails? This is also an issue with locks though, if it is not cleanned up, but with locks we can safely clean up if it fails, compared to clean up a pending that might delete a concurrent put result.

We don't need concurrent requests, we just want one per CID.

Reply:
i realized by having DENYLIST we avoid the need for a seperate lock entry.

Reply:
The concurrent issue won't be in GET route (where we changed to only check DENYLIST). It will happen for GET (List after refactor we talked), PUT, DELETE of same key in concurrent POST actions